Convert STM32 CAN driver to DT macros usage

erwango opened the pull request:
Converting stm32 CAN driver to new DT api.

Tricky part is the fact that can2 requires access to can1 reg value, while it should be allowed to be enabled while keeping can1 disabled (besides using both instances at same time is not possible).

In a first attempt, I solved this by enabling both instances and adding a runtime check to ensure only one instance was in use at a time. Which means to use can2, can1 need to be enabled, which is sub-optimal regarding footprint consumption.

Second attempt added an optional property to provide access to can1 from can2, by adding an optional master-can prop pointing to can1 as phandle.. which requires can1 to be enabled.

This last attempt use an optional prop which is directly the value of the master can register. So can2 can be enabled w/o enabling can1.

Raised https://github.com/zephyrproject-rtos/zephyr/issues/24686 in order to go back to second proposal as soon as allowed by DT api

erwango committed:
drivers/can: stm32: configure driver thanks to compatible

Remove need to configure from soc/

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32: Replace CAN Kconfig symbols by dt instances

Replace use of CAN_X Kconfig symbols by access to dt instances

Control of enabled instances is now done using dts nodes status
field ('okay'/'disabled'). For boards supporting only can2,
can1 instance is disabled.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak reviewed (commented):
can you add a build check that validates the DT_NODELABEL(can1) / DT_NODELABEL(can2) are DT_NODE_HAS_COMPAT(st_stm32_can).

I think I want any drivers that use NODELABEL to validate the compat is what they think it is (also helps match driver to binding).

erwango commented:
> can you add a build check that validates the DT_NODELABEL(can1) / DT_NODELABEL(can2) are DT_NODE_HAS_COMPAT(st_stm32_can).
> 

Ok, updated.

> I think I want any drivers that use NODELABEL to validate the compat is what they think it is (also helps match driver to binding).

Then , maybe we should provide a macro to do that then, like:
```
#define DT_HAS_COMPAT_NODELABEL_(device, compat) \
        DT_HAS_NODE(DT_NODELABEL(device)) && \
	DT_NODE_HAS_COMPAT(DT_NODELABEL(device), compat)
```

We could also reuse `DT_DRV_COMPAT`. So that it becomes a general rule that driver  define `DT_DRV_COMPAT`. And have a macro like:
```
#define DT_HAS_NODELABEL_(device) \
        DT_HAS_NODE(DT_NODELABEL(device)) && \
	DT_NODE_HAS_COMPAT(DT_NODELABEL(device), DT_DRV_COMPAT)
```

In any case, it this becomes a rule, I think we need a ready to use macro.

erwango committed:
drivers/can: stm32: Convert to DT macros

Use DT macros to configure stm32 can device instances.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on drivers/can/can_stm32.c:
```diff
@@ -1126,47 +1139,52 @@ NET_DEVICE_INIT(socket_can_stm32_1, SOCKET_CAN_NAME_1, socket_can_init_1,
-	.prop_ts1 = DT_CAN_2_PROP_SEG + DT_CAN_2_PHASE_SEG1,
-	.ts2 = DT_CAN_2_PHASE_SEG2,
+	.can = (CAN_TypeDef *)DT_REG_ADDR(DT_NODELABEL(can2)),
+	.master_can = (CAN_TypeDef *)DT_REG_ADDR(DT_NODELABEL(can1)),
```
@alexanderwachter, @galak this is the main impact of the change.
To use dma2, we now need to compile dma1.
To handle this, I added 3e77fe090828b2b760ad509aad702c897fc03c56

alexanderwachter replied:
That's no the only impact. We are also wasting memory because we instantiate both deriver instances. The reason why we have the CAN_2 is because of pin multiplexing. But on Slack somebody announced to implement support for both CAN simultaneously. Maybe we can find another solution to reference to CAN_1. Maybe with a phandle?

erwango replied:
> We are also wasting memory because we instantiate both driver instances.

Of course

> Maybe we can find another solution to reference to CAN_1. Maybe with a phandle?

Maybe it can be done through dedicated prop on can2 description, something like master-can, or whatever. This would be optional as can1 would not need it. Can you check this particular point?

erwango replied:
Edited PR description on various attempts to solve this.

erwango committed:
dts/bindings: st,stm32-can: Add optional master-can property

In STM32 can driver CAN2 instance need to know the base address
of CAN1 instance which is referenced as master can.
Add this information as an optional property to stm32 can binding.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Setting DNM as I pushed to share non working work

galak commented on dts/bindings/can/st,stm32-can.yaml:
```diff
@@ -13,3 +13,8 @@ properties:
     clocks:
       required: true
+
+    master-can-reg:
```
I'm ok with this, but can we open an issue to fix this in the future to just be a phandle to the master node.

erwango replied:
#24686

alexanderwachter reviewed (commented):
LGTM. I‘ll run the tests on my boards ASAP.
